File: sim/control/orbit/predictive_engagement.py
"""Deterministic short-horizon pursuit and evasion policies for RPO.

The policies in this module use HCW as an onboard prediction model only.  The
authoritative trajectory continues to be propagated by the scenario's normal
orbit dynamics through the flight-software actuation path.
"""

from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from math import isfinite

import numpy as np

from sim.dynamics.orbit.relative_linear import RelativeLinearDynamics

MAX_PREDICTION_STEPS = 10_000
MAX_ACCELERATION_FRACTIONS = 8
MAX_PREDICTIVE_PROPAGATIONS = 2_000_000


@dataclass(frozen=True, slots=True)
class PredictiveAction:
    acceleration_ric_m_s2: tuple[float, float, float]
    predicted_closest_range_m: float
    predicted_closest_time_s: float
    predicted_capture_time_s: float | None
    phase: str


def select_intercept_action(
    state_ric_si: np.ndarray,
    *,
    mean_motion_rad_s: float,
    max_acceleration_m_s2: float,
    horizon_s: float,
    step_s: float,
    pulse_duration_s: float,
    capture_radius_m: float,
    capture_margin_m: float = 0.0,
    acceleration_fractions: tuple[float, ...] = (0.5, 1.0),
    target_acceleration_ric_m_s2: np.ndarray | None = None,
) -> PredictiveAction:
    """Select a coast-aware pursuer action against one target action.

    Coast is selected whenever the passive trajectory reaches the capture
    radius.  Otherwise the lowest-delta-v candidate that predicts capture is
    used; if none captures, the candidate with the best closest approach is
    selected.
    """

    state = _validated_state(state_ric_si)
    _validate_policy_inputs(
        mean_motion_rad_s=mean_motion_rad_s,
        max_acceleration_m_s2=max_acceleration_m_s2,
        horizon_s=horizon_s,
        step_s=step_s,
        pulse_duration_s=pulse_duration_s,
        capture_radius_m=capture_radius_m,
        capture_margin_m=capture_margin_m,
        acceleration_fractions=acceleration_fractions,
    )
    target_action = (
        np.zeros(3)
        if target_acceleration_ric_m_s2 is None
        else np.asarray(target_acceleration_ric_m_s2, dtype=float).reshape(3)
    )
    if not np.all(np.isfinite(target_action)):
        raise ValueError("target_acceleration_ric_m_s2 must be finite")

    actions = _candidate_actions(state, max_acceleration_m_s2, acceleration_fractions)
    _require_work_budget(len(actions), horizon_s=horizon_s, step_s=step_s)
    predictions = [
        _predict(
            state,
            action - target_action,
            mean_motion_rad_s=mean_motion_rad_s,
            horizon_s=horizon_s,
            step_s=step_s,
            pulse_duration_s=pulse_duration_s,
            capture_radius_m=capture_radius_m,
        )
        for action in actions
    ]
    passive = predictions[0]
    passive_threshold = max(capture_radius_m - capture_margin_m, 0.0)
    if passive.closest_range_m <= passive_threshold:
        return _result(actions[0], passive, "passive_intercept_coast")

    capturing = [
        (index, prediction)
        for index, prediction in enumerate(predictions)
        if prediction.capture_time_s is not None
    ]
    if capturing:
        index, prediction = min(
            capturing,
            key=lambda item: (
                float(np.linalg.norm(actions[item[0]])) * pulse_duration_s,
                float(item[1].capture_time_s or 0.0),
                item[1].closest_range_m,
            ),
        )
        phase = "intercept_burn" if np.linalg.norm(actions[index]) > 0.0 else "intercept_coast"
        return _result(actions[index], prediction, phase)

    index, prediction = min(
        enumerate(predictions),
        key=lambda item: (
            item[1].closest_range_m,
            item[1].closest_time_s,
            float(np.linalg.norm(actions[item[0]])) * pulse_duration_s,
        ),
    )
    phase = "intercept_search_burn" if np.linalg.norm(actions[index]) > 0.0 else "intercept_search_coast"
    return _result(actions[index], prediction, phase)


def select_evasion_action(
    state_ric_si: np.ndarray,
    *,
    mean_motion_rad_s: float,
    max_acceleration_m_s2: float,
    opponent_max_acceleration_m_s2: float,
    horizon_s: float,
    step_s: float,
    pulse_duration_s: float,
    capture_radius_m: float,
    capture_margin_m: float = 0.0,
    acceleration_fractions: tuple[float, ...] = (0.5, 1.0),
) -> PredictiveAction:
    """Choose an evasion action against the pursuer's best bounded response.

    For every target candidate the pursuer is allowed to select its own best
    action.  The target maximizes the resulting worst-case closest approach,
    preferring survival, later capture, and lower delta-v in that order.
    """

    state = _validated_state(state_ric_si)
    _validate_policy_inputs(
        mean_motion_rad_s=mean_motion_rad_s,
        max_acceleration_m_s2=max_acceleration_m_s2,
        horizon_s=horizon_s,
        step_s=step_s,
        pulse_duration_s=pulse_duration_s,
        capture_radius_m=capture_radius_m,
        capture_margin_m=capture_margin_m,
        acceleration_fractions=acceleration_fractions,
    )
    if not isfinite(opponent_max_acceleration_m_s2) or opponent_max_acceleration_m_s2 < 0.0:
        raise ValueError("opponent_max_acceleration_m_s2 must be finite and nonnegative")

    target_actions = _candidate_actions(state, max_acceleration_m_s2, acceleration_fractions)
    pursuer_actions = _candidate_actions(-state, opponent_max_acceleration_m_s2, acceleration_fractions)
    _require_work_budget(
        len(target_actions) * len(pursuer_actions),
        horizon_s=horizon_s,
        step_s=step_s,
    )
    outcomes: list[tuple[int, _Prediction]] = []
    for target_index, target_action in enumerate(target_actions):
        responses = [
            _predict(
                state,
                target_action - pursuer_action,
                mean_motion_rad_s=mean_motion_rad_s,
                horizon_s=horizon_s,
                step_s=step_s,
                pulse_duration_s=pulse_duration_s,
                capture_radius_m=capture_radius_m,
            )
            for pursuer_action in pursuer_actions
        ]
        pursuer_best = min(responses, key=_pursuer_outcome_key)
        outcomes.append((target_index, pursuer_best))

    target_index, prediction = max(
        outcomes,
        key=lambda item: _evader_outcome_key(
            item[1],
            action=target_actions[item[0]],
            pulse_duration_s=pulse_duration_s,
        ),
    )
    action = target_actions[target_index]
    phase = "predictive_evasion_burn" if np.linalg.norm(action) > 0.0 else "predictive_evasion_coast"
    return _result(action, prediction, phase)


@dataclass(frozen=True, slots=True)
class _Prediction:
    closest_range_m: float
    closest_time_s: float
    capture_time_s: float | None


def _predict(
    state: np.ndarray,
    relative_acceleration: np.ndarray,
    *,
    mean_motion_rad_s: float,
    horizon_s: float,
    step_s: float,
    pulse_duration_s: float,
    capture_radius_m: float,
) -> _Prediction:
    current = np.asarray(state, dtype=float).copy()
    elapsed = 0.0
    closest_range = float(np.linalg.norm(current[:3]))
    closest_time = 0.0
    capture_time = 0.0 if closest_range <= capture_radius_m else None
    while elapsed < horizon_s - 1.0e-12:
        dt = min(step_s, horizon_s - elapsed)
        if elapsed < pulse_duration_s < elapsed + dt:
            dt = pulse_duration_s - elapsed
        ad, bd = _discrete_matrices(float(mean_motion_rad_s), float(dt))
        action = relative_acceleration if elapsed < pulse_duration_s - 1.0e-12 else np.zeros(3)
        current = ad @ current + bd @ action
        elapsed += dt
        range_m = float(np.linalg.norm(current[:3]))
        if range_m < closest_range:
            closest_range = range_m
            closest_time = elapsed
        if capture_time is None and range_m <= capture_radius_m:
            capture_time = elapsed
    return _Prediction(closest_range, closest_time, capture_time)


@lru_cache(maxsize=64)
def _discrete_matrices(mean_motion_rad_s: float, dt_s: float) -> tuple[np.ndarray, np.ndarray]:
    return RelativeLinearDynamics.hcw(mean_motion_rad_s).discrete_matrices(dt_s)


def _candidate_actions(
    state: np.ndarray,
    maximum: float,
    fractions: tuple[float, ...],
) -> tuple[np.ndarray, ...]:
    actions: list[np.ndarray] = [np.zeros(3)]
    if maximum <= 0.0:
        return tuple(actions)
    directions = [
        np.array((1.0, 0.0, 0.0)),
        np.array((-1.0, 0.0, 0.0)),
        np.array((0.0, 1.0, 0.0)),
        np.array((0.0, -1.0, 0.0)),
        np.array((0.0, 0.0, 1.0)),
        np.array((0.0, 0.0, -1.0)),
    ]
    for vector in (state[:3], state[3:], state[:3] + state[3:] * 120.0):
        norm = float(np.linalg.norm(vector))
        if norm > 1.0e-12:
            unit = vector / norm
            directions.extend((unit, -unit))
    unique: list[np.ndarray] = []
    for direction in directions:
        if not any(float(np.dot(direction, existing)) > 1.0 - 1.0e-10 for existing in unique):
            unique.append(direction)
    for fraction in fractions:
        actions.extend(maximum * float(fraction) * direction for direction in unique)
    return tuple(actions)


def _pursuer_outcome_key(prediction: _Prediction) -> tuple[float, float, float]:
    captures = prediction.capture_time_s is not None
    return (
        0.0 if captures else 1.0,
        float(prediction.capture_time_s) if captures else prediction.closest_range_m,
        prediction.closest_range_m,
    )


def _evader_outcome_key(
    prediction: _Prediction,
    *,
    action: np.ndarray,
    pulse_duration_s: float,
) -> tuple[float, float, float, float]:
    survives = prediction.capture_time_s is None
    return (
        1.0 if survives else 0.0,
        prediction.closest_range_m if survives else float(prediction.capture_time_s or 0.0),
        prediction.closest_time_s,
        -float(np.linalg.norm(action)) * pulse_duration_s,
    )


def _result(action: np.ndarray, prediction: _Prediction, phase: str) -> PredictiveAction:
    return PredictiveAction(
        tuple(float(value) for value in action),
        prediction.closest_range_m,
        prediction.closest_time_s,
        prediction.capture_time_s,
        phase,
    )


def _validated_state(value: np.ndarray) -> np.ndarray:
    state = np.asarray(value, dtype=float).reshape(6)
    if not np.all(np.isfinite(state)):
        raise ValueError("state_ric_si must contain six finite values")
    return state


def _validate_policy_inputs(
    *,
    mean_motion_rad_s: float,
    max_acceleration_m_s2: float,
    horizon_s: float,
    step_s: float,
    pulse_duration_s: float,
    capture_radius_m: float,
    capture_margin_m: float,
    acceleration_fractions: tuple[float, ...],
) -> None:
    for name, value, positive in (
        ("mean_motion_rad_s", mean_motion_rad_s, True),
        ("max_acceleration_m_s2", max_acceleration_m_s2, False),
        ("horizon_s", horizon_s, True),
        ("step_s", step_s, True),
        ("pulse_duration_s", pulse_duration_s, True),
        ("capture_radius_m", capture_radius_m, True),
        ("capture_margin_m", capture_margin_m, False),
    ):
        if not isfinite(value) or (value <= 0.0 if positive else value < 0.0):
            qualifier = "positive" if positive else "nonnegative"
            raise ValueError(f"{name} must be finite and {qualifier}")
    if step_s > pulse_duration_s:
        raise ValueError("step_s must be no greater than pulse_duration_s")
    if pulse_duration_s > horizon_s:
        raise ValueError("pulse_duration_s must be no greater than horizon_s")
    if capture_margin_m >= capture_radius_m:
        raise ValueError("capture_margin_m must be smaller than capture_radius_m")
    if not acceleration_fractions or any(
        not isfinite(value) or value <= 0.0 or value > 1.0 for value in acceleration_fractions
    ):
        raise ValueError("acceleration_fractions must contain values in (0, 1]")
    if len(acceleration_fractions) > MAX_ACCELERATION_FRACTIONS:
        raise ValueError(
            f"acceleration_fractions may contain at most {MAX_ACCELERATION_FRACTIONS} values"
        )
    steps = int(np.ceil(horizon_s / step_s)) + 1
    if steps > MAX_PREDICTION_STEPS:
        raise ValueError(f"predictive horizon requires more than {MAX_PREDICTION_STEPS} propagation steps")


def _require_work_budget(candidate_count: int, *, horizon_s: float, step_s: float) -> None:
    steps = int(np.ceil(float(horizon_s) / float(step_s))) + 1
    requested = int(candidate_count) * steps
    if requested > MAX_PREDICTIVE_PROPAGATIONS:
        raise ValueError(
            "predictive candidate search exceeds the bounded propagation budget "
            f"({requested} > {MAX_PREDICTIVE_PROPAGATIONS})"
        )

File: web/rpo-duel-prototype/src/shared/predictive-engagement.js
export const PREDICTIVE_ENGAGEMENT_POLICY = deepFreeze({
  schema_version: "rpo-duel.computer-policy.v1",
  prediction_model: "HCW",
  horizon_s: 1800,
  step_s: 30,
  pulse_duration_s: 60,
  decision_interval_s: 120,
  capture_margin_m: 20,
  acceleration_fractions: [0.5, 1],
});

export function selectInterceptAction(stateRicSi, options = {}) {
  const state = validateState(stateRicSi);
  const config = validateOptions(options);
  const targetAction = options.target_acceleration_ric_m_s2 === undefined
    ? [0, 0, 0]
    : validateVector(options.target_acceleration_ric_m_s2, "target_acceleration_ric_m_s2");
  const actions = candidateActions(state, config.max_acceleration_m_s2, config.acceleration_fractions);
  const predictions = actions.map((action) => predict(
    state,
    subtract(action, targetAction),
    config,
  ));

  const passive = predictions[0];
  const passiveThreshold = Math.max(config.capture_radius_m - config.capture_margin_m, 0);
  if (passive.closest_range_m <= passiveThreshold) {
    return result(actions[0], passive, "passive_intercept_coast");
  }

  let capturing = null;
  for (let index = 0; index < predictions.length; index += 1) {
    const prediction = predictions[index];
    if (prediction.capture_time_s === null) continue;
    const key = [
      norm(actions[index]) * config.pulse_duration_s,
      prediction.capture_time_s || 0,
      prediction.closest_range_m,
    ];
    if (!capturing || compareKeys(key, capturing.key) < 0) capturing = { index, prediction, key };
  }
  if (capturing) {
    const action = actions[capturing.index];
    const phase = norm(action) > 0 ? "intercept_burn" : "intercept_coast";
    return result(action, capturing.prediction, phase);
  }

  let best = null;
  for (let index = 0; index < predictions.length; index += 1) {
    const prediction = predictions[index];
    const key = [
      prediction.closest_range_m,
      prediction.closest_time_s,
      norm(actions[index]) * config.pulse_duration_s,
    ];
    if (!best || compareKeys(key, best.key) < 0) best = { index, prediction, key };
  }
  const action = actions[best.index];
  const phase = norm(action) > 0 ? "intercept_search_burn" : "intercept_search_coast";
  return result(action, best.prediction, phase);
}

export function selectEvasionAction(stateRicSi, options = {}) {
  const state = validateState(stateRicSi);
  const config = validateOptions(options);
  const opponentMaximum = nonnegative(
    options.opponent_max_acceleration_m_s2,
    "opponent_max_acceleration_m_s2",
  );
  const targetActions = candidateActions(state, config.max_acceleration_m_s2, config.acceleration_fractions);
  const pursuerActions = candidateActions(
    state.map((value) => -value),
    opponentMaximum,
    config.acceleration_fractions,
  );

  let targetBest = null;
  for (let targetIndex = 0; targetIndex < targetActions.length; targetIndex += 1) {
    const targetAction = targetActions[targetIndex];
    let pursuerBest = null;
    for (const pursuerAction of pursuerActions) {
      const prediction = predict(state, subtract(targetAction, pursuerAction), config);
      const key = pursuerOutcomeKey(prediction);
      if (!pursuerBest || compareKeys(key, pursuerBest.key) < 0) pursuerBest = { prediction, key };
    }
    const key = evaderOutcomeKey(pursuerBest.prediction, targetAction, config.pulse_duration_s);
    if (!targetBest || compareKeys(key, targetBest.key) > 0) {
      targetBest = { targetIndex, prediction: pursuerBest.prediction, key };
    }
  }

  const action = targetActions[targetBest.targetIndex];
  const phase = norm(action) > 0 ? "predictive_evasion_burn" : "predictive_evasion_coast";
  return result(action, targetBest.prediction, phase);
}

function predict(state, relativeAcceleration, config) {
  let current = state.slice();
  let elapsed = 0;
  let closestRange = norm(current.slice(0, 3));
  let closestTime = 0;
  let captureTime = closestRange <= config.capture_radius_m ? 0 : null;
  while (elapsed < config.horizon_s - 1.0e-12) {
    let dt = Math.min(config.step_s, config.horizon_s - elapsed);
    if (elapsed < config.pulse_duration_s && config.pulse_duration_s < elapsed + dt) {
      dt = config.pulse_duration_s - elapsed;
    }
    const action = elapsed < config.pulse_duration_s - 1.0e-12
      ? relativeAcceleration
      : ZERO_VECTOR;
    current = propagateHcwConstantAcceleration(
      current,
      action,
      config.mean_motion_rad_s,
      dt,
    );
    elapsed += dt;
    const range = norm(current.slice(0, 3));
    if (range < closestRange) {
      closestRange = range;
      closestTime = elapsed;
    }
    if (captureTime === null && range <= config.capture_radius_m) captureTime = elapsed;
  }
  return {
    closest_range_m: closestRange,
    closest_time_s: closestTime,
    capture_time_s: captureTime,
  };
}

export function propagateHcwConstantAcceleration(stateRicSi, accelerationRicMps2, meanMotionRadS, dtS) {
  const state = validateState(stateRicSi);
  const acceleration = validateVector(accelerationRicMps2, "acceleration_ric_m_s2");
  const n = positive(meanMotionRadS, "mean_motion_rad_s");
  const dt = positive(dtS, "dt_s");
  const nt = n * dt;
  const cosine = Math.cos(nt);
  const sine = Math.sin(nt);
  const n2 = n * n;
  const [r, i, c, rd, id, cd] = state;
  const [ar, ai, ac] = acceleration;

  return [
    (4 - 3 * cosine) * r + sine / n * rd + 2 * (1 - cosine) / n * id
      + (1 - cosine) / n2 * ar + 2 * (nt - sine) / n2 * ai,
    6 * (sine - nt) * r + i - 2 * (1 - cosine) / n * rd + (4 * sine - 3 * nt) / n * id
      - 2 * (nt - sine) / n2 * ar + (4 * (1 - cosine) / n2 - 1.5 * dt * dt) * ai,
    cosine * c + sine / n * cd + (1 - cosine) / n2 * ac,
    3 * n * sine * r + cosine * rd + 2 * sine * id
      + sine / n * ar + 2 * (1 - cosine) / n * ai,
    -6 * n * (1 - cosine) * r - 2 * sine * rd + (4 * cosine - 3) * id
      - 2 * (1 - cosine) / n * ar + (4 * sine / n - 3 * dt) * ai,
    -n * sine * c + cosine * cd + sine / n * ac,
  ];
}

function candidateActions(state, maximum, fractions) {
  const actions = [[0, 0, 0]];
  if (maximum <= 0) return actions;
  const directions = [
    [1, 0, 0], [-1, 0, 0], [0, 1, 0], [0, -1, 0], [0, 0, 1], [0, 0, -1],
  ];
  for (const vector of [
    state.slice(0, 3),
    state.slice(3, 6),
    state.slice(0, 3).map((value, index) => value + state[index + 3] * 120),
  ]) {
    const magnitude = norm(vector);
    if (magnitude > 1.0e-12) {
      const unit = vector.map((value) => value / magnitude);
      directions.push(unit, unit.map((value) => -value));
    }
  }
  const unique = [];
  for (const direction of directions) {
    if (!unique.some((existing) => dot(direction, existing) > 1 - 1.0e-10)) unique.push(direction);
  }
  for (const fraction of fractions) {
    for (const direction of unique) {
      actions.push(direction.map((value) => maximum * fraction * value));
    }
  }
  return actions;
}

function pursuerOutcomeKey(prediction) {
  const captures = prediction.capture_time_s !== null;
  return [
    captures ? 0 : 1,
    captures ? prediction.capture_time_s : prediction.closest_range_m,
    prediction.closest_range_m,
  ];
}

function evaderOutcomeKey(prediction, action, pulseDurationS) {
  const survives = prediction.capture_time_s === null;
  return [
    survives ? 1 : 0,
    survives ? prediction.closest_range_m : prediction.capture_time_s || 0,
    prediction.closest_time_s,
    -norm(action) * pulseDurationS,
  ];
}

function result(action, prediction, phase) {
  return {
    acceleration_ric_m_s2: action.slice(),
    predicted_closest_range_m: prediction.closest_range_m,
    predicted_closest_time_s: prediction.closest_time_s,
    predicted_capture_time_s: prediction.capture_time_s,
    phase,
  };
}

function validateOptions(options) {
  const fractions = Array.from(options.acceleration_fractions || []);
  const config = {
    mean_motion_rad_s: positive(options.mean_motion_rad_s, "mean_motion_rad_s"),
    max_acceleration_m_s2: nonnegative(options.max_acceleration_m_s2, "max_acceleration_m_s2"),
    horizon_s: positive(options.horizon_s, "horizon_s"),
    step_s: positive(options.step_s, "step_s"),
    pulse_duration_s: positive(options.pulse_duration_s, "pulse_duration_s"),
    capture_radius_m: positive(options.capture_radius_m, "capture_radius_m"),
    capture_margin_m: nonnegative(options.capture_margin_m ?? 0, "capture_margin_m"),
    acceleration_fractions: fractions.map((value) => positive(value, "acceleration_fractions")),
  };
  if (config.step_s > config.pulse_duration_s) throw new Error("step_s must be no greater than pulse_duration_s");
  if (config.pulse_duration_s > config.horizon_s) throw new Error("pulse_duration_s must be no greater than horizon_s");
  if (config.capture_margin_m >= config.capture_radius_m) {
    throw new Error("capture_margin_m must be smaller than capture_radius_m");
  }
  if (!fractions.length || config.acceleration_fractions.some((value) => value > 1)) {
    throw new Error("acceleration_fractions must contain values in (0, 1]");
  }
  return config;
}

function validateState(value) {
  const state = Array.from(value || [], Number);
  if (state.length !== 6 || state.some((item) => !Number.isFinite(item))) {
    throw new Error("state_ric_si must contain six finite values");
  }
  return state;
}

function validateVector(value, name) {
  const vector = Array.from(value || [], Number);
  if (vector.length !== 3 || vector.some((item) => !Number.isFinite(item))) {
    throw new Error(`${name} must contain three finite values`);
  }
  return vector;
}

function positive(value, name) {
  const number = Number(value);
  if (!Number.isFinite(number) || number <= 0) throw new Error(`${name} must be finite and positive`);
  return number;
}

function nonnegative(value, name) {
  const number = Number(value);
  if (!Number.isFinite(number) || number < 0) throw new Error(`${name} must be finite and nonnegative`);
  return number;
}

function compareKeys(first, second) {
  for (let index = 0; index < first.length; index += 1) {
    if (first[index] < second[index]) return -1;
    if (first[index] > second[index]) return 1;
  }
  return 0;
}

function subtract(first, second) {
  return first.map((value, index) => value - second[index]);
}

function dot(first, second) {
  return first.reduce((total, value, index) => total + value * second[index], 0);
}

function norm(values) {
  return Math.hypot(...values);
}

function deepFreeze(value) {
  if (!value || typeof value !== "object" || Object.isFrozen(value)) return value;
  Object.freeze(value);
  Object.values(value).forEach(deepFreeze);
  return value;
}

const ZERO_VECTOR = Object.freeze([0, 0, 0]);
